Documentation/Evolution/CharacterClasses.md

```diff
@@ -0,0 +1,354 @@
+// true
+```
+
+Instead of treating `“e”` and `"\u{301}"` as separate elements, Swift’s string recognizes the pair as a grapheme cluster, as represented by the `Character` type, and as canonically equivalent to their single Unicode scalar value counterpart. Characters that are made up of multiple Unicode scalars include emoji (flags are usually 2 Unicode scalar values, family groupings can be nearly a dozen), Korean hangeul (1-3 Unicode scalar values per character), and Thai script, among many others.
+
+This is a departure from many other languages and frameworks, such as Python and Foundation's NSString, where individual Unicode scalars, or sometimes their UTF-16 or UTF-8 encoding units, are treated as the elements of a string.
+
+```python
+for ch in u'Cafe\u0301':
+    print("# " + ch)
+# C
+# a
+# f
+# e
+#´
+```
+
+Just as Python strings are different from Swift's, in these other languages, regular expressions behave in ways that do not match the expectations of a Swift string user.
+
+Character classes in other languages match at the Unicode scalar (or encoded unit) value, instead of recognizing grapheme clusters as characters. When matching the `.` character class, other languages will only match the first part of an `"e\u{301}"` grapheme cluster. Some languages, like Perl, Ruby, and Java, support an additional `\X` metacharacter, which explictly represents a single grapheme cluster.
+
+| Matching  `"Cafe\u{301}"` | Pattern: `^Caf.` | Remaining | Pattern:  `^Caf\X` | Remaining |
+|---|---|---|---|---|
+| NSString, C#, Rust, Go | `"Cafe"` | `"´"` | n/a | n/a |
+| Java, Ruby, Perl | `"Cafe"` | `"´"` | `"Café"` | `""` |
+
+By and large, other languages do not use canonical equivalence when matching characters in regular expressions. Java provides a `CANON_EQ` option when compiling a pattern to opt into such checking.
+
+```java
+String cafe = "Cafe\u0301";
+
+Pattern exactPatt = Pattern.compile("é");
+exactPatt.matcher(cafe).find()      // false
+
+Pattern equivPatt = Pattern.compile("é", Pattern.CANON_EQ);
+equivPatt.matcher(cafe).find()      // true
+```
+
+Finally, there is wide variance of the actual definitions of character classes between different languages. As one example, should the `\d` character class, which matches digits, include the digits from the _Halfwidth and Fullwidth Forms_ Unicode block? ICU-based regular expression engines, like those used in Java and Foundation's `NSRegularExpression`, include these extra digits, while other languages do not. Determining exactly what characters are matched by each character class of a particular language or library can be difficult, making it harder to fully understand what kinds of strings a pattern will or will not match.
+
+## Proposed Solution
+
+We propose a treatment of character classes in Swift regular expression that is consistent with the Swift string model, configurable at the point of declaration, and comprehensible for users getting to know Swift regular expressions.
+
```

Comment:
What about configurable at point of use? Or configurable by API? I don't think we can make this decision here, as we don't have any existing pitched API to reference, nor are we pitching any such API here.

This also buries the lead and the point of this pitch. Support pattern matching over strings, including regex, requires defining character classes. We have some prior art in Character properties and Unicode Scalar properties, but those are incomplete for matching. Let's beef them up, clarifying what they correspond to in regex literals (and using regex as a forcing function for this much needed semantic work).

There's so much semantic work that we can focus on that instead of where mode is configured, while acknowledging that configuring the mode is a big unknown and pending the next step and investigation. Either way, we want to flesh out this semantic space so we can make concrete progress here.